### PR TITLE
[red-knot] Integrate formatter

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This is a first hacky approach to integrate the formatter into red-knot. 

There are plenty of TODOs. The main questions is where and when we should write the formatted content and how we avoid that formatting requires two-passes for the cache to be "hot". 

## Test Plan

I ran red knot and it formated my files (all lints gone, whoops)

### CODE REVIEW

File: crates/red_knot/src/main.rs [hunk 1]
```diff
@@ -138,22 +140,28 @@ impl MainLoop {
...
+                                    })
+                                    .unwrap();
+                            }
+                            Err(QueryError::Cancelled) => {}
+                        });
+                    }
+
+                    if !program.is_cancelled() {
+                        let _ = program.format();
+                    }
```
Comment (MichaReiser) [inline crates/red_knot/src/main.rs:164]: That's a bit hacked in. We can also decide to remove it for now. But it does show that formatting after linting is possible (but requires a `&mut Program`)

File: crates/red_knot/src/format.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    // TODO use the `format_module` method here to re-use the AST.
+
+    let result =
+        ruff_python_formatter::format_range(source.text(), range, PyFormatOptions::default())?;
+    Ok(result)
+}
+
+/// Checks if the file is correctly formatted. It creates a diagnostic for formatting issues.
+#[tracing::instrument(level = "trace", skip(db))]
+pub(crate) fn check_formatted<Db>(db: &Db, file_id: FileId) -> Result<Diagnostics, FormatError>
```
Comment (carljm) [inline crates/red_knot/src/format.rs:74]: for consistency?

```suggestion
pub(crate) fn check_file_formatted<Db>(db: &Db, file_id: FileId) -> Result<Diagnostics, FormatError>
```

File: crates/red_knot/src/program/format.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    pub fn format(&mut self) -> QueryResult<()> {
+        // Formats all open files
+
+        // TODO make `Executor` from `check` reusable.
+        for file in self.workspace.open_files() {
+            match self.format_file(file) {
+                Ok(FormattedFile::Formatted(content)) => {
+                    let path = self.file_path(file);
+
+                    // TODO: This is problematic because it immediately re-triggers the file watcher.
```
Comment (carljm) [inline crates/red_knot/src/program/format.rs:16]: It seems like mostly this is a good thing? The file content has changed, the rest of the system should react accordingly, as it normally would. I don't think we should special-case re-formatting in the rest of our invalidation -- it's a file change like any other.

The only place where this seems undesirable is that you would like to track that a file that we just formatted is still formatted; we don't have to reformat it. So I would scope any special-casing to this specific point only, contained within the formatter logic. Maybe we could keep a cache mapping the hash of known-well-formatted contents for a file, so when we get updated sources for a file from the file watcher, we can check if the hash of the new contents matches an entry in this cache, and if so, directly mark it as already well-formatted?

File: crates/red_knot/src/program/mod.rs [hunk 4]
```diff
@@ -124,6 +132,24 @@ impl LintDb for Program {
...
+    fn format_file_range(
+        &self,
+        file_id: FileId,
+        range: TextRange,
+    ) -> Result<PrintedRange, FormatError> {
+        format_file_range(self, file_id, range)
+    }
+
+    fn check_file_formatted(&self, file_id: FileId) -> Result<Diagnostics, FormatError> {
+        check_formatted(self, file_id)
```
Comment (carljm) [inline crates/red_knot/src/program/mod.rs:149]:
```suggestion
        check_file_formatted(self, file_id)
```

**Review by carljm — APPROVED** [review]

**Comment by MichaReiser** [comment]
I still plan to land this PR eventually but aren't prioritising it right now. Deciding on a query system and exploring persistent caching is of higher importance right now.

**Pull request closed without merging**